service/storage_proxy.cc

```diff
@@ -3723,23 +3739,26 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
             });
         }
     }
-    future<rpc::tuple<query::result_digest, api::timestamp_type, cache_temperature>> make_digest_request(gms::inet_address ep, clock_type::time_point timeout) {
+    future<rpc::tuple<query::result_digest, api::timestamp_type, cache_temperature, rpc::optional<full_position>>> make_digest_request(gms::inet_address ep, clock_type::time_point timeout) {
         ++_proxy->get_stats().digest_read_attempts.get_ep_stat(get_topology(), ep);
         if (fbu::is_me(ep)) {
             tracing::trace(_trace_state, "read_digest: querying locally");
             return _proxy->query_result_local_digest(_schema, _cmd, _partition_range, _trace_state,
-                        timeout, digest_algorithm(*_proxy), adjust_rate_limit_for_local_operation(_rate_limit_info));
+                        timeout, digest_algorithm(*_proxy), adjust_rate_limit_for_local_operation(_rate_limit_info)).then([] (rpc::tuple<query::result_digest, api::timestamp_type, cache_temperature, full_position>&& r) {
+                return rpc::tuple<query::result_digest, api::timestamp_type, cache_temperature, rpc::optional<full_position>>(std::move(r));
+            });
```

Comment:
This introduces an allocation. Maybe we can change query_result_local_digest() to return the expected type.

Reply:
I want to express in the signature of `query_result_local_digest()` that the position is *always* provided by the replica. The `rpc::optional<>` is only added for the RPC layer, to allow for talking to older nodes.

Reply:
BTW, why does this introduce an allocation? I use the move constructor of `rpc::tuple`. Is it because of the additional task?

Reply:
Yes, the extra task.

Reply:
> I want to express in the signature of `query_result_local_digest()` that the position is _always_ provided by the replica. The `rpc::optional<>` is only added for the RPC layer, to allow for talking to older nodes.

I agree with that in principle (of course). But this is too much on the hot path and will definitely be measurable (it won't be dominant, but it will be a small regression). I think this case merits breaking the rules. We can run perf_simple_query to verify it.

Reply:
Ok, I changed it.